Ci add luacheck

carlsmedstad opened the pull request:
Resolves #1253.

I tried to do this as non-intrusively as possible since I'm not very familiar with the code base. Hopefully, the fixes and ignores are considered appropriate. If not, I'm very open to reworking them with feedback from the maintainers.

carlsmedstad commented on src/luarocks/tools/zip.lua:
```diff
@@ -10,7 +10,7 @@ local dir = require("luarocks.dir")
 local pack = table.pack or function(...) return { n = select("#", ...), ... } end
 
-local stat_ok, stat = pcall(require, "posix.sys.stat")
+pcall(require, "posix.sys.stat")
```
I assumed so as well, but was not certain, so I left it. I'll remove it and see if it breaks any tests.

hishamhm commented:
Hi there! Thank you for contributing! :)

I've never integrated luacheck in CI before because I already have it integrated into my editor, so most of the serious things that luacheck catches I tend to spot immediately, and there were a couple of things that luacheck does warn about that I never found important enough (I've recently noticed that I had never merged in my `.luacheckrc` after all these years, so hopefully that gives an idea of the kinds of errors I usually ignore). Plus, luacheck is currently in need of a maintainer, so I've been reluctant on adding it as a dependency.

Still, I see this PR has caught a couple of mistakes here and there, which is nice!

One change I find concerning, though, is the many variables renamed to `_`. Given a dynamically typed code-base, having the proper names serves as useful documentation to what the functions return, even if we're not using currently them right now. I wonder if [this feature](https://luacheck.readthedocs.io/en/stable/warnings.html#secondary-values-and-variables) could silence those warnings and keep those variables properly named?

carlsmedstad commented:
Hey! :wave:

My pleasure, I'm very appreciative of Lua having such good package management tooling so I'm glad to be able to contribute :) 

Your suggestion of using the option `unused_secondaries` did indeed remove the warnings. I've updated the PR and squashed the commits.

hishamhm commented:
@carlsmedstad Thank you for the adjustments! It's now merged :)

hishamhm commented:
I guess I'll add a luacheck step to my personal git pre-commit hook now!

hishamhm commented:
```
$ cat .git/hooks/pre-commit
#!/bin/sh

luacheck --cache .luacheckcache -q src/
```
